llvm/lib/Target/X86/X86TargetTransformInfo.cpp

```diff
@@ -1559,6 +1559,22 @@ InstructionCost X86TTIImpl::getShuffleCost(
       return TTI::TCC_Free;
   }
 
+  // Attempt to detect a cheaper inlane shuffle, avoiding 128-bit subvector
+  // permutation.
+  bool IsInLaneShuffle = false;
+  if (BaseTp->getPrimitiveSizeInBits() > 0 &&
+      (BaseTp->getPrimitiveSizeInBits() % 128) == 0 &&
+      Mask.size() == BaseTp->getElementCount().getKnownMinValue()) {
```

Comment:
Shall we also check that the original type of the vector element is the same as the type of the "legalized" (machine) vector type?

Reply:
Sure, I think checking the scalar sizes match should be enough - I'll add that shortly